### ISSUE TITLE
Hash function improvements

### ISSUE THREAD

**Issue by Riolku** [kuzudb/kuzu#2290]
Currently, our hash function is quite naive, and it's very easy to cause very bad collisions in the hash index.

We should replace it. DuckDB seems to be using a very simple but efficient hash function, which seems to be a good choice based on some cursory reading online.

We should also change the name since we aren't _actually_ using murmurhash.

**Comment by Riolku** [comment]
Changing the integer hash is straightforward, but there are some major remaining questions:

1. What do we do about our CombineHash function? DuckDB uses a simple XOR, but this means `ab` and `ba` have the same hash, which is generally undesirable, or so I've been told. 
2. Semi-related, our string hash algorithm is `std::hash`. Should we change this? DuckDB and Postgres are both using something super specialized.

**Comment by benjaminwinger** [comment]
> Semi-related, our string hash algorithm is std::hash. Should we change this? DuckDB and Postgres are both using something super specialized.

Simple answer is yes: #2952.
DuckDB is using a string function from [here](https://github.com/martinus/robin-hood-hashing), which is actually just [MurmurHash2](https://github.com/rurban/smhasher/blob/master/MurmurHash2.cpp). I did a quick benchmark and the performance is almost identical to the rolling hash implementation I wrote for #2952; though I don't know if it produces better results. Postgres seems to be using a hash function from [here](https://burtleburtle.net/bob/hash/doobs.html); the author says that murmurhash is faster, but weaker, than his hash function ([the postgres version](https://github.com/postgres/postgres/blob/6979ea2638a51c40acf6d04b816550b2c35b3e55/src/common/hashfn.c#L37) has an update from 2006, but I think we'd be better off looking at newer hash function comparisons like [smhasher](https://github.com/rurban/smhasher) if we want to find something better).

Also, I think this issue was already more or less implemented in 359c09d5ad, so maybe this could be closed after #2952 is merged.

**Comment by Riolku** [comment]
Yes, integer hashing was changed by me. String hashing, seems like you have a PR for. I think the other hash functions are defined in terms of those two.
